integration/spark/shared/src/main/java/io/openlineage/spark/agent/lifecycle/plan/LogicalRDDVisitor.java

```diff
@@ -27,21 +29,27 @@
 public class LogicalRDDVisitor<D extends OpenLineage.Dataset>
     extends AbstractRDDNodeVisitor<LogicalRDD, D> {
 
+  Set<RDD<?>> flattenedRdds;
```

Comment:
So far we were able to keep visitors stateless. Keeping rdds reference won't release references to rdds until the next RDD operation is performed. 

Does it make sense in this case to compute `flattenedRdds` twice (within `isDefinedAt` and `apply` method? What's the reason to change `isDefinedAt` method? I don't understand how does this relate to infinite loop?

Reply:
this change is not related to the infinite loop , but just an observation that flattening was called repeatedly and that can be avoided. the call that is repeated in the is definedAt and apply is the Rdds.findFileLikeRdds , which in theory can be called once as well / cached. I however limited to only saving the results of the flatten call , which I have found to be expensive when the RDD tree is large .